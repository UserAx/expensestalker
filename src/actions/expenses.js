import database from '../firebase/firebase';

// ADD_EXPENSE
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

export const startAddExpense = (expenseData ={}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      description = '',
      note='',
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = {description, note, amount, createdAt};
    return database.ref(`users/${uid}/expenses`).push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    });
  }
}

export const startSetExpenses = ( ) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses`).once('value').then((snapshot) => {
      const expenses = [];
      snapshot.forEach((childSnapshot) => {
        expenses.push({id: childSnapshot.key,
                      ...childSnapshot.val()});
      });
      dispatch(setExpenses(expenses));
    });
  }
}

export const startRemoveExpense = ({id}={}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses/${id}`).remove().then(()=> {
      dispatch(removeExpense({id}));
    });
  }
}

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({//Note here we pass id as {id: 'someId'}
  type: 'REMOVE_EXPENSE',
  id
});

export const startEditExpense = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses/${id}`).update(updates).then(()=> {
      dispatch(editExpense(id, updates));
    });
  }
}


// EDIT_EXPENSE
export const editExpense = (id, updates) => ({//However here we pass argument as ('someId', {description: 'Some Text'})
  type: 'EDIT_EXPENSE',
  id,
  updates
});