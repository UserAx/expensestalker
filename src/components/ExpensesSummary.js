import React from 'react';
import {Link} from 'react-router-dom';
import {connect} from 'react-redux';
import numeral from 'numeral';
import selectExpensesTotal from '../selectors/expenses-total';
import selectExpenses from '../selectors/expenses';

export const ExpensesSummary = ({expensesCount, expensesTotal}) => {
    const expenseWord = expensesCount === 1 ? 'expense' : 'expenses';
    const formatedTotals = numeral(expensesTotal/100).format('$0,0.00');
    return (
        <div className="page-header">
            <div className="content-container">
                <h3 className="page-header-title">Viewing <span>{expensesCount}</span> {expenseWord}, 
                a total of <span>{formatedTotals}</span>. </h3>
                <div className="page-header__actions">
                    <Link className="button" to="/create">Add Expense</Link>
                </div>
            </div>
        </div>
    );
};

//we need this because our expenses exist in the redux store state object.
const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);
    return {
        expensesCount: visibleExpenses.length,
        expensesTotal: selectExpensesTotal(visibleExpenses)
    };
};

export default connect(mapStateToProps)(ExpensesSummary);