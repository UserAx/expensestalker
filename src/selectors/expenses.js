// Get visible expenses
import moment from 'moment';

export default (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses.filter((expense) => {//returns only those expense object that match the following cases.
    const createdAtMoment = moment(expense.createdAt);
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
    //Here startDate is getting compared to createdAtMoment not the other war around. 
    //if start day is yesterday and createdAtMoment is yesterday then we get true since startday is before today.
    //Same is true for my endDateMatch bellow.
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

    return startDateMatch && endDateMatch && textMatch;
  }).sort((a, b) => {//and sorts those object by amount or date if set.
    if (sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1;
    } else if (sortBy === 'amount') {
      return a.amount < b.amount ? 1 : -1;
    }
  });
};