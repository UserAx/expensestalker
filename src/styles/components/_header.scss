.header {
    background: $dark-blue;
}

.header__content {
    align-items: center;
    display: flex;
    justify-content: space-between;
    padding: $s-size 0;
}

.header__title{
    color: white;
    text-decoration: none;
    h1 {
        margin: 0
    }
}

.content-container{
    margin: 0 auto;
    padding: 0 $m-size;
    max-width: 80rem;
}