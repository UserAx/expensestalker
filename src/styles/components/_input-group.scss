.input-group {
    display: flex;
    margin-bottom: $m-size;
    flex-direction: column;
    @media (min-width: $desktop-breakpoint) {//when the minimum size becomes the one of a desktop
        flex-direction: row;
        margin-bottom: $l-size;
    }
}

.input-group__item{
    margin: 0 $s-size 0 0;
    @media (min-width: $desktop-breakpoint) {
        margin-bottom: $s-size;
    }
}