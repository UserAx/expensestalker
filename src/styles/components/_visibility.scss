.show-for-mobile {
    @media(min-width: $desktop-breakpoint) {//when the minimum size becomes the one of a desktop
        display: none;
    }
}

.show-for-desktop {
    @media(max-width: $desktop-breakpoint - .01rem) {//when the maximum size becomes the one of a mobile
        display: none;
    }
}