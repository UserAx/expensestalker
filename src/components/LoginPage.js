import React from 'react';
import {startLogin} from '../actions/auth';
import {connect } from 'react-redux';

export const LoginPage = ({startLogin}) => (
    <div className ="box-layout">
        <div className="box-layout__box">
            <h1 className="box-layout__title">Expensify App</h1>
            <p>It is time to get your expenses under control.</p>
            <button className="button" onClick={startLogin}>Log In</button>
        </div>
    </div>
);

const mapDispatchToProps = (dispatch) => ({
    startLogin: () => dispatch(startLogin())
});

export default connect(undefined, mapDispatchToProps)(LoginPage);