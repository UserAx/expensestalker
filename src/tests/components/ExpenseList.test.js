import React from 'react';
import {shallow} from 'enzyme';
import {ExpenseList} from '../../components/ExpenseList';

import expenses from '../fixtures/expenses';

test('should render ExpenseList with expenses', () => {
    const wrapper = shallow(<ExpenseList expenses = {expenses.map((expense)=> ({...expense, id:'staticIdforTesting'}))} />);
    //I have set a static id because every time the test runs the id is going to be unique i.e. it will fail every second time.
    expect(wrapper).toMatchSnapshot();
});

test('should render ExpenseList with no expenses provided', () => {
    const wrapper = shallow(<ExpenseList expenses = {[]} />);
    expect(wrapper).toMatchSnapshot();
});
