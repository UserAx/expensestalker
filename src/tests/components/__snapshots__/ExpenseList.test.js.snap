// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with expenses 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Expense
    </div>
    <div
      className="show-for-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <ExpenseListItem
      amount={1500}
      createdAt={0}
      description="rent"
      id="staticIdforTesting"
      key="staticIdforTesting"
      note="This month's rent."
    />
    <ExpenseListItem
      amount={150}
      createdAt={-345600000}
      description="Shoping"
      id="staticIdforTesting"
      key="staticIdforTesting"
      note="Brought eggs."
    />
    <ExpenseListItem
      amount={2000}
      createdAt={345600000}
      description="Gas Bill"
      id="staticIdforTesting"
      key="staticIdforTesting"
      note="This month's gas bill."
    />
  </div>
</div>
`;

exports[`should render ExpenseList with no expenses provided 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Expense
    </div>
    <div
      className="show-for-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <div
      className="list-item list-item--message"
    >
      <span>
        No Expenses
      </span>
    </div>
  </div>
</div>
`;
